### PR TITLE
[NEW] Add the ability to set the character message limit on Livechat widget

### PR DESCRIPTION
**Pull request by oguhpereira**
## Proposed changes

Implementing a new character limit field in the LiveChat Aparence.

## Issue(s)

#17451

## How to test or reproduce

1. Open Livechat Aparence
2. Edit Characther Limit input
3. Save
4. reload the page and check if the information has been saved

If you want test with Livechat, you will need this [feature](https://github.com/RocketChat/Rocket.Chat.Livechat/pull/443)

## Screenshots

image

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

## Changelog

additional setting in Admin panel for setting a character limit on livechat.

### CODE REVIEW

File: app/livechat/client/views/app/livechatAppearance.js [hunk 1]
```diff
@@ -150,6 +155,7 @@ Template.livechatAppearance.onCreated(async function() {
 
 	this.title.set(livechatTitle && livechatTitle.value);
 	this.color.set(livechatTitleColor && livechatTitleColor.value);
+	this.limitTextLength.set(livechatTitleLimitTextLength && livechatTitleLimitTextLength.value);
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:158]:
```suggestion
	this.limitTextLength.set(livechatMessageCharacterLimit && livechatMessageCharacterLimit.value);
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 2]
```diff
@@ -187,6 +193,9 @@ Template.livechatAppearance.events({
 		const settingTitleColor = getSettingFromAppearance(instance, 'Livechat_title_color');
 		instance.color.set(settingTitleColor && settingTitleColor.value);
 
+		const settingTitleLimitTextLength = getSettingFromAppearance(instance, 'Livechat_title_character_limit');
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:196]:
```suggestion
		const settinglivechatMessageCharacterLimit = getSettingFromAppearance(instance, 'Livechat_message_character_limit');
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 3]
```diff
@@ -187,6 +193,9 @@ Template.livechatAppearance.events({
...
+		instance.limitTextLength.set(settingTitleLimitTextLength && settingTitleLimitTextLength.value);
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:197]:
```suggestion
		instance.limitTextLength.set(settinglivechatMessageCharacterLimit && settinglivechatMessageCharacterLimit.value);
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 4]
```diff
@@ -240,6 +249,10 @@ Template.livechatAppearance.events({
 				_id: 'Livechat_title_color',
 				value: instance.color.get(),
 			},
+			{
+				_id: 'Livechat_title_character_limit',
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:253]:
```suggestion
				_id: 'Livechat_message_character_limit',
```

File: app/livechat/server/api/lib/appearance.js [hunk 5]
```diff
@@ -10,6 +10,7 @@ export async function findAppearance({ userId }) {
 			$in: [
 				'Livechat_title',
 				'Livechat_title_color',
+				'Livechat_title_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/api/lib/appearance.js:13]:
```suggestion
				'Livechat_message_character_limit',
```

File: app/livechat/server/api/lib/livechat.js [hunk 6]
```diff
@@ -111,6 +111,7 @@ export function settings() {
 			forceAcceptDataProcessingConsent: initSettings.Livechat_force_accept_data_processing_consent,
 			showConnecting: initSettings.Livechat_Show_Connecting,
 			agentHiddenInfo: initSettings.Livechat_show_agent_info === false,
+			limitTextLength: parseInt(initSettings.Livechat_title_character_limit),
```
Comment (renatobecker) [inline app/livechat/server/api/lib/livechat.js:114]:
```suggestion
			limitTextLength: parseInt(initSettings.Livechat_message_character_limit),
```

File: app/livechat/server/config.js [hunk 7]
```diff
@@ -27,6 +27,13 @@ Meteor.startup(function() {
 		public: true,
 	});
 
+	settings.add('Livechat_title_character_limit', null, {
```
Comment (renatobecker) [inline app/livechat/server/config.js:30]:
```suggestion
	settings.add('Livechat_message_character_limit', 0, {
```

File: app/livechat/server/lib/Livechat.js [hunk 8]
```diff
@@ -428,6 +428,7 @@ export const Livechat = {
 		Settings.findNotHiddenPublic([
 			'Livechat_title',
 			'Livechat_title_color',
+			'Livechat_title_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/lib/Livechat.js:431]:
```suggestion
			'Livechat_message_character_limit',
```

File: app/livechat/server/methods/getInitialData.js [hunk 9]
```diff
@@ -66,6 +66,7 @@ Meteor.methods({
 
 		info.title = initSettings.Livechat_title;
 		info.color = initSettings.Livechat_title_color;
+		info.limitTextLength = initSettings.Livechat_title_character_limit;
```
Comment (renatobecker) [inline app/livechat/server/methods/getInitialData.js:69]: This method is no longer used, so you don't need to change it.

File: app/livechat/server/methods/saveAppearance.js [hunk 10]
```diff
@@ -12,6 +12,7 @@ Meteor.methods({
 		const validSettings = [
 			'Livechat_title',
 			'Livechat_title_color',
+			'Livechat_title_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/methods/saveAppearance.js:15]:
```suggestion
			'Livechat_message_character_limit',
```

File: app/livechat/client/views/app/livechatAppearance.html [hunk 11]
```diff
@@ -16,6 +16,10 @@ <h2>{{_ "Settings"}}</h2>
 						<label for="color">{{_ "Title_bar_color"}}</label>
 						<input type="color" class="preview-settings rc-input__element" name="color" id="color" value="{{color}}" />
 					</div>
+					<div class="input-line">
+						<label for="limitTextLength">{{_ "Message_Characther_Limit"}}</label>
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.html:20]: This translation is still missing.

**Comment by oguhpereira** [comment]
@renatobecker I added a new option that allows to enable or disable the display of the character limit

image

It has also been configured so that it fetches the maximum message size setting. The Message Characther Limit field will be used to set a size smaller than the default for the maximum message size setting.

File: packages/rocketchat-i18n/i18n/en.i18n.json [hunk 12]
```diff
@@ -2272,6 +2272,8 @@
   "Livechat_Take_Confirm": "Do you want to take this client?",
   "Livechat_title": "Livechat Title",
   "Livechat_title_color": "Livechat Title Background Color",
+  "Livechat_show_message_character_limit": "Livechat message character limit",
```
Comment (renatobecker) [inline packages/rocketchat-i18n/i18n/en.i18n.json:2275]:
```suggestion
  "Livechat_enable_message_character_limit": "Enable message character limit",
```

File: packages/rocketchat-i18n/i18n/pt-BR.i18n.json [hunk 13]
```diff
@@ -2016,6 +2016,8 @@
   "Livechat_Take_Confirm": "Você quer levar esse cliente?",
   "Livechat_title": "Título Livechat",
   "Livechat_title_color": "Cor de fundo do título do Livechat",
+  "Livechat_show_message_character_limit": "Exibir Limite de caracteres da mensagem no livechat",
```
Comment (renatobecker) [inline packages/rocketchat-i18n/i18n/pt-BR.i18n.json:2019]:
```suggestion
  "Livechat_enable_message_character_limit": "Habilitar Limite de caracteres de mensagem",
```

File: packages/rocketchat-i18n/i18n/pt.i18n.json [hunk 14]
```diff
@@ -1894,6 +1894,8 @@
   "Livechat_Take_Confirm": "Quer levar este cliente?",
   "Livechat_title": "Título Livechat",
   "Livechat_title_color": "Cor de fundo do título do Livechat",
+  "Livechat_show_message_character_limit": "Exibir Limite de caracteres da mensagem no livechat",
```
Comment (renatobecker) [inline packages/rocketchat-i18n/i18n/pt.i18n.json:1897]:
```suggestion
  "Livechat_enable_message_character_limit": "Habilitar Limite de caracteres de mensagem",
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 15]
```diff
@@ -240,6 +257,14 @@ Template.livechatAppearance.events({
 				_id: 'Livechat_title_color',
 				value: instance.color.get(),
 			},
+			{
+				_id: 'Livechat_show_message_character_limit',
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:261]:
```suggestion
				_id: 'Livechat_enable_message_character_limit',
```

File: app/livechat/server/lib/Livechat.js [hunk 16]
```diff
@@ -428,6 +428,8 @@ export const Livechat = {
 		Settings.findNotHiddenPublic([
 			'Livechat_title',
 			'Livechat_title_color',
+			'Livechat_show_message_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/lib/Livechat.js:431]:
```suggestion
			'Livechat_enable_message_character_limit',
```

File: app/livechat/server/methods/saveAppearance.js [hunk 17]
```diff
@@ -12,6 +12,8 @@ Meteor.methods({
 		const validSettings = [
 			'Livechat_title',
 			'Livechat_title_color',
+			'Livechat_show_message_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/methods/saveAppearance.js:15]:
```suggestion
			'Livechat_enable_message_character_limit',
```

File: app/livechat/client/views/app/livechatAppearance.html [hunk 18]
```diff
@@ -16,6 +16,19 @@ <h2>{{_ "Settings"}}</h2>
 						<label for="color">{{_ "Title_bar_color"}}</label>
 						<input type="color" class="preview-settings rc-input__element" name="color" id="color" value="{{color}}" />
 					</div>
+					<div class="input-line">
+						<label for="showLimitTextLength">{{_ "Livechat_show_message_character_limit"}}</label>
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.html:20]:
```suggestion
						<label for="showLimitTextLength">{{_ "Livechat_enable_message_character_limit"}}</label>
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 19]
```diff
@@ -132,6 +139,8 @@ Template.livechatAppearance.onCreated(async function() {
 
 	const livechatTitle = getSettingFromAppearance(this, 'Livechat_title');
 	const livechatTitleColor = getSettingFromAppearance(this, 'Livechat_title_color');
+	const livechatShowMessageCharacterLimit = getSettingFromAppearance(this, 'Livechat_show_message_character_limit');
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:142]:
```suggestion
	const livechatShowMessageCharacterLimit = getSettingFromAppearance(this, 'Livechat_enable_message_character_limit');
```

File: app/livechat/client/views/app/livechatAppearance.js [hunk 20]
```diff
@@ -187,6 +198,12 @@ Template.livechatAppearance.events({
 		const settingTitleColor = getSettingFromAppearance(instance, 'Livechat_title_color');
 		instance.color.set(settingTitleColor && settingTitleColor.value);
 
+		const settinglivechatShowMessageCharacterLimit = getSettingFromAppearance(instance, 'Livechat_show_message_character_limit');
```
Comment (renatobecker) [inline app/livechat/client/views/app/livechatAppearance.js:201]:
```suggestion
		const settinglivechatShowMessageCharacterLimit = getSettingFromAppearance(instance, 'Livechat_enable_message_character_limit');
```

File: app/livechat/server/config.js [hunk 21]
```diff
@@ -27,6 +27,20 @@ Meteor.startup(function() {
 		public: true,
 	});
 
+	settings.add('Livechat_show_message_character_limit', false, {
```
Comment (renatobecker) [inline app/livechat/server/config.js:30]:
```suggestion
	settings.add('Livechat_enable_message_character_limit', false, {
```

File: app/livechat/server/lib/Livechat.js [hunk 22]
```diff
@@ -428,6 +428,8 @@ export const Livechat = {
...
+			'Livechat_message_character_limit',
```
Comment (renatobecker) [inline app/livechat/server/lib/Livechat.js:432]: You need to retrieve here the `Message_MaxAllowedSize` setting as well.
```suggestion
			'Livechat_message_character_limit',
			'Message_MaxAllowedSize',
```

File: app/livechat/server/api/lib/livechat.js [hunk 23]
```diff
@@ -66,6 +67,19 @@ export function findOpenRoom(token, departmentId) {
 	return room;
 }
 
+export function findLimitTextLength({ Livechat_show_message_character_limit, Livechat_message_character_limit}) {
```
Comment (renatobecker) [inline app/livechat/server/api/lib/livechat.js:70]: You don't need this function.

File: app/livechat/server/api/lib/livechat.js [hunk 24]
```diff
@@ -94,6 +108,7 @@ export function settings() {
 	const departments = findDepartments();
 	const sound = `${ Meteor.absoluteUrl() }sounds/chime.mp3`;
 	const emojis = Meteor.call('listEmojiCustom');
+	const limitTextLength = findLimitTextLength(initSettings);
```
Comment (renatobecker) [inline app/livechat/server/api/lib/livechat.js:111]: You don't need this const.

File: app/livechat/server/api/lib/livechat.js [hunk 25]
```diff
@@ -111,6 +126,7 @@ export function settings() {
 			forceAcceptDataProcessingConsent: initSettings.Livechat_force_accept_data_processing_consent,
 			showConnecting: initSettings.Livechat_Show_Connecting,
 			agentHiddenInfo: initSettings.Livechat_show_agent_info === false,
+			limitTextLength,
```
Comment (renatobecker) [inline app/livechat/server/api/lib/livechat.js:129]:
```suggestion
			limitTextLength: initSettings.Livechat_enabled_message_character_limit
				&& (initSettings.Livechat_message_character_limit || initSettings.messageMaxAllowedSize),
```

File: app/livechat/server/config.js [hunk 26]
```diff
@@ -27,6 +27,20 @@ Meteor.startup(function() {
...
+		type: 'int',
```
Comment (renatobecker) [inline app/livechat/server/config.js:31]:
```suggestion
		type: 'boolean',
```

**Comment by oguhpereira** [comment]
@renatobecker I changed the review requests but I was in doubt if it was to change the following items:

* Name `show`LimitTextLength have to change for `enable`LimitTextLength in functions and variables? 
* In this [function](https://github.com/RocketChat/Rocket.Chat/pull/18261#discussion_r463920586) there was a condition that when the Livechat_message_character_limit is greater than Message_MaxAllowedSize. Is required this condition?

File: packages/rocketchat-i18n/i18n/pt-BR.i18n.json [hunk 27]
```diff
@@ -2168,6 +2170,7 @@
   "Message_AudioRecorderEnabled_Description": "Requer arquivos 'audio / mp3' para ser um tipo de mídia aceito dentro das configurações 'Upload de arquivo'.",
   "Message_BadWordsFilterList": "Adicione palavrões para a lista negra",
   "Message_BadWordsFilterListDescription": "Adicione palavrões separados por vírgula para filtrar",
+  "Message_Characther_Limit": "Limite do caráter da mensagem",
```
Comment (dlassalle) [inline packages/rocketchat-i18n/i18n/pt-BR.i18n.json:2173]: FYI `Characther` -> `Character`, but its the same everywhere for this so it still functions fine

Comment (renatobecker) [reply packages/rocketchat-i18n/i18n/pt-BR.i18n.json:2173]: Thanks, @dlassalle. I'll fix this typo.

**Pull request merged**